op-chain-ops/deployer/pipeline/opchain.go

```diff
@@ -91,13 +94,46 @@ func DeployOPChain(ctx context.Context, env *Env, artifactsFS foundry.StatDirFs,
 		DelayedWETHPermissionedGameProxyAddress:   dco.DelayedWETHPermissionedGameProxy,
 		DelayedWETHPermissionlessGameProxyAddress: dco.DelayedWETHPermissionlessGameProxy,
 	})
+
+	currentBlock, _ := env.L1Client.BlockNumber(ctx)
+	block, _ := env.L1Client.BlockByNumber(ctx, big.NewInt(int64(currentBlock)))
+	currentBlockHash := block.Hash()
+
+	// If any of the implementations addresses (excluding OpcmProxy) are empty,
+	// we need to set them using the addresses of the corresponding proxies.
+	// The reason these might be empty is because we're only invoking DeployOPChain.s.sol as part of the pipeline.
+	// TODO: Need to initialize 'mipsSingletonAddress' and 'preimageOracleSingletonAddress'
+	setEIP1967ImplementationAddress(ctx, env.L1Client, dco.DelayedWETHPermissionedGameProxy, currentBlockHash, &st.ImplementationsDeployment.DelayedWETHImplAddress)
+	setEIP1967ImplementationAddress(ctx, env.L1Client, dco.OptimismPortalProxy, currentBlockHash, &st.ImplementationsDeployment.OptimismPortalImplAddress)
+	setEIP1967ImplementationAddress(ctx, env.L1Client, dco.SystemConfigProxy, currentBlockHash, &st.ImplementationsDeployment.SystemConfigImplAddress)
+	setRDPImplementationAddress(ctx, env.L1Client, dco.AddressManager, &st.ImplementationsDeployment.L1CrossDomainMessengerImplAddress)
+	setEIP1967ImplementationAddress(ctx, env.L1Client, dco.L1ERC721BridgeProxy, currentBlockHash, &st.ImplementationsDeployment.L1ERC721BridgeImplAddress)
+	setEIP1967ImplementationAddress(ctx, env.L1Client, dco.L1StandardBridgeProxy, currentBlockHash, &st.ImplementationsDeployment.L1StandardBridgeImplAddress)
+	setEIP1967ImplementationAddress(ctx, env.L1Client, dco.OptimismMintableERC20FactoryProxy, currentBlockHash, &st.ImplementationsDeployment.OptimismMintableERC20FactoryImplAddress)
+	setEIP1967ImplementationAddress(ctx, env.L1Client, dco.DisputeGameFactoryProxy, currentBlockHash, &st.ImplementationsDeployment.DisputeGameFactoryImplAddress)
```

Comment:
I doubt this is the main contributor to runtime, but I wonder if we can dispatch these all in parallel or use JSON RPC batching to improve performance? Can be done in a separate PR, not a blocker

Reply:
https://github.com/ethereum-optimism/optimism/pull/12434/commits/b117cf4c00602f6a753d03eb88b1c2a7d1fcc2b3

Parallel execution added